implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,237 @@
+package io.smallrye.config;
+
+import static io.smallrye.common.classloader.ClassPathUtils.consumeAsPath;
+import static io.smallrye.common.classloader.ClassPathUtils.consumeAsPaths;
+import static io.smallrye.config.Converters.newCollectionConverter;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.eclipse.microprofile.config.spi.Converter;
+
+import io.smallrye.common.annotation.Experimental;
+
+/**
+ * This {@code AbstractLocationConfigSourceFactory} allows to initialize additional config locations with the
+ * configuration {@link AbstractLocationConfigSourceFactory#SMALLRYE_LOCATIONS}.
+ * <p>
+ *
+ * Locations set in {@link AbstractLocationConfigSourceFactory#SMALLRYE_LOCATIONS} are loaded in order and from the
+ * following resources:
+ *
+ * <ol>
+ * <li>file or directory</li>
+ * <li>classpath resource</li>
+ * <li>jar resource</li>
+ * <li>http resource</li>
+ * </ol>
+ * <p>
+ *
+ * If a profile is active, the profile resource is loaded if available and the unprofiled resource is also
+ * available. This is to keep a consistent order with the unprofiled resource. Profiles are not taken into account if
+ * the location is a directory.
+ */
+@Experimental("Load additional config locations")
+public abstract class AbstractLocationConfigSourceFactory implements ConfigSourceFactory {
+    public static final String SMALLRYE_LOCATIONS = "smallrye.config.locations";
+
+    protected abstract String[] getFileExtensions();
+
+    protected abstract ConfigSource loadConfigSource(final URL url, final int ordinal) throws IOException;
+
+    protected ConfigSource loadConfigSource(final URL url) throws IOException {
+        return this.loadConfigSource(url, ConfigSource.DEFAULT_ORDINAL);
+    }
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context) {
```

Comment:
When did we make the switch to returning multiple config sources here?  It means that it is impossible to assemble a chain in priority order because the config sources would have to be reshuffled by priority after the chain is built - meaning that each factory will ultimately not be "looking at" lower ordinal config sources as they are being built, which was a big part of the intent of the suggestion.  If this was my idea, then slap me. :-)

Reply:
We ended up changing the priority on the factory to be used only for initialization purposes. The main reason is to be able to honour the `config_ordinal` property if set in the `ConfigSource`. So, yes we do initialize with the initial chain and then reshuffle all the sources in the right order.